presto-main/src/main/java/io/prestosql/operator/scalar/StringFunctions.java

```diff
@@ -877,6 +880,74 @@ public static Slice toUtf8(@SqlType("varchar(x)") Slice slice)
         return slice;
     }
 
+    private static final String UTF8_DEFAULT_REPLACEMENT = "\uFFFD";
+
+    private static String stringFromCharset(Slice charset, Slice slice)
+    {
+        try {
+            return slice.toString(Charset.forName(charset.toStringUtf8()));
+        }
+        catch (UnsupportedCharsetException e) {
+            throw new PrestoException(INVALID_FUNCTION_ARGUMENT, "Unsupported character set");
+        }
+        catch (IllegalCharsetNameException e) {
+            throw new PrestoException(INVALID_FUNCTION_ARGUMENT, "Illegal character set");
+        }
+    }
+
+    @Description("Decodes string with a specific character set")
+    @ScalarFunction
+    @SqlType(StandardTypes.VARCHAR)
+    public static Slice fromCharset(@SqlType("varchar") Slice charset, @SqlType(StandardTypes.VARBINARY) Slice slice, @SqlType(StandardTypes.BIGINT) long replacementCodePoint)
+    {
+        if (replacementCodePoint > MAX_CODE_POINT || Character.getType((int) replacementCodePoint) == SURROGATE) {
+            throw new PrestoException(INVALID_FUNCTION_ARGUMENT, "Invalid replacement character");
+        }
+
+        String stringFromCharset = stringFromCharset(charset, slice);
+        Slice replacementCharacter = chr(replacementCodePoint);
+
+        return Slices.utf8Slice(stringFromCharset.replaceAll(UTF8_DEFAULT_REPLACEMENT, replacementCharacter.toStringUtf8()));
+    }
+
+    @Description("Decodes string with a specific character set")
+    @ScalarFunction
+    @SqlType(StandardTypes.VARCHAR)
+    public static Slice fromCharset(@SqlType("varchar") Slice charset, @SqlType(StandardTypes.VARBINARY) Slice slice, @SqlType("varchar") Slice replacementCharacter)
+    {
+        int count = countCodePoints(replacementCharacter);
+        if (count > 1) {
+            throw new PrestoException(INVALID_FUNCTION_ARGUMENT, "Replacement character string must empty or a single character");
+        }
+
+        String stringFromCharset = stringFromCharset(charset, slice);
+        return Slices.utf8Slice(stringFromCharset.replaceAll(UTF8_DEFAULT_REPLACEMENT, replacementCharacter.toStringUtf8()));
```

Comment:
Based on the discussion above, we can't rely on the default replacement character. Instead, we should use `CharsetDecoder` directly via `charset.newDecoder()` to do the conversion. You can use `replaceWith()` to set the replacement character. For an empty replacement, you'll have to use `onMalformedInput(IGNORE)` instead since `CharsetDecoder` doesn't allow an empty replacement.

Make sure to add test cases for replacement with different values and with an empty string.